Add approve/revoke delegate section to Token Program #51 (#139)
Unboxed-Software/solana-course: content/intro-to-reading-data.md

- **Accounts** are like the files in Solana’s network ledger. All state data is stored in an account. Accounts can be used for many things, but for now we’ll focus on the aspect of accounts which store SOL.
- **SOL** is the name of Solana’s native token.
- **Lamports** are fractional SOL and are named after [Leslie Lamport](https://en.wikipedia.org/wiki/Leslie_Lamport). 1 Lamport = 0.000000001 SOL.
- **Public keys**, often referred to as addresses, point to accounts on the Solana network. While you must have a specific secret key to perform certain functions within accounts, anyone can read account data with a public key.
- **JSON RPC API**: all interactions with the Solana network happens through the [JSON RPC API](https://docs.solana.com/developing/clients/jsonrpc-api). This is effectively an HTTP POST with a JSON body that represents the method you want to call.

> - **@solana/web3.js** is an abstraction on top of the JSON RPC API. It can be installed with `npm` or `yarn` and allows you to call Solana JSON RPC API as JavaScript functions. For example, you can use it to query the SOL balance of any account:

Reviewer: `or yarn` isn't necessary. We don't need to name every alternative package manager, and Anchor etc are removing the yarn dependency.

> - **@solana/web3.js** is an abstraction on top of the JSON RPC API. It can be installed with `npm` or `yarn` and allows you to call Solana JSON RPC API as JavaScript functions. For example, you can use it to query the SOL balance of any account:

Reviewer: ```suggestion
- **@solana/web3.js** an `npm` module that invokes the Solana JSON RPC API through JavaScript functions. For example, to query the SOL balance of an account:
```

>     ```tsx

Reviewer: ```suggestion
    ```typescript
```

    import {
      clusterApiUrl,
      Connection,
      PublicKey,
      LAMPORTS_PER_SOL,
    } from "@solana/web3.js";

    async function getBalanceUsingWeb3(address: PublicKey): Promise<number> {
        const connection = new Connection(clusterApiUrl('devnet'), 'confirmed');
        return connection.getBalance(address);
    }

    const publicKey = new PublicKey('7C4jsPZpht42Tw6MjXWF56Q5RQUocjBBmciEjDa8HRtp');
    getBalanceUsingWeb3(publicKey).then(balance => {
        console.log(`${balance/LAMPORTS_PER_SOL} SOL`)

>     });

Reviewer: Not your fault (this is old code), but the formatting is odd here - can you check they're using 2-spaces per CONTRIBUTING? I normally just cut and paste into a text file, save and let prettier do it's work, then paste back.

### Public Keys

> Public keys are often referred to as addresses. The addresses point to accounts on the Solana network. If you want to run a specific program a.k.a smart contract or just want to transfer some SOL, you’ll need to provide the necessary public key (or keys) to do so.

Reviewer: 'Smart contract' is not used on Solana. https://solana.com/docs/terminology#onchain-program

> Public keys are often referred to as addresses. The addresses point to accounts on the Solana network. If you want to run a specific program a.k.a smart contract or just want to transfer some SOL, you’ll need to provide the necessary public key (or keys) to do so.

Reviewer: ```suggestion
Public keys are used as addresses for accounts on the Solana network. When you run a function in a specific onchain program, Solana will contact the program at it's address, and run the associated instruction handler. For example, to transfer some SOL, you run the `SystemAccount` program's `transfer` instruction handler. 
```

Reviewer: We can probably do better than this, that's just a suggestion.

> Public keys are 32-bytes and they are often shown as [base-58](https://en.bitcoinwiki.org/wiki/Base58) encoded strings like `7C4jsPZpht42Tw6MjXWF56Q5RQUocjBBmciEjDa8HRtp`.

Reviewer: I like this, 32 bytes is probably more practical. Also the link is valuable, most people won't know base58.

> *JSON-RPC is a stateless, light-weight remote procedure call (RPC) protocol. Primarily this specification defines several data structures and the rules around their processing. It is transport agnostic in that the concepts can be used within the same process, over sockets, over http, or in many various message passing environments. It uses [JSON](http://www.json.org/) ([RFC 4627](http://www.ietf.org/rfc/rfc4627.txt)) as data format.*
>

> In practice, this specification simply involves sending a JSON object representing a method you want to call. You can do this with [websocket](https://en.wikipedia.org/wiki/WebSocket), http, and more.

Reviewer: ```suggestion
In practice, this specification simply involves sending a JSON object representing a method you want to call. You can do this with [websockets](https://en.wikipedia.org/wiki/WebSocket), http, and more.
```

This JSON object needs four members:

> - `jsonrpc` - The JSON RPC version number. This needs to be *exactly* `&quot;2.0&quot;`.

Reviewer: HTML quotes are no longer necessary since UTF 8 happened about 20 years ago.
